File: src/main/java/AoC2016_21.java
import static com.github.pareronia.aoc.CharArrayUtils.indexOf;
import static com.github.pareronia.aoc.CharArrayUtils.reverse;
import static java.lang.Boolean.FALSE;
import static java.lang.Boolean.TRUE;

import java.util.Arrays;
import java.util.List;

import com.github.pareronia.aoc.StringOps;
import com.github.pareronia.aoc.solution.SolutionBase;

@SuppressWarnings({"PMD.NoPackage", "PMD.ClassNamingConventions"})
public final class AoC2016_21 extends SolutionBase<List<String>, String, String> {

    private AoC2016_21(final boolean debug) {
        super(debug);
    }

    public static AoC2016_21 create() {
        return new AoC2016_21(false);
    }

    public static AoC2016_21 createDebug() {
        return new AoC2016_21(true);
    }

    @Override
    protected List<String> parseInput(final List<String> inputs) {
        return inputs;
    }

    private String solve(final List<String> operations, final String start, final boolean encrypt) {
        char[] ch = start.toCharArray();
        for (int i = 0; i < operations.size(); i++) {
            final int idx = encrypt ? i : operations.size() - 1 - i;
            final String operation = operations.get(idx);
            if (operation.startsWith("swap position ")) {
                final Integer[] params = StringOps.getDigits(operation, 2);
                final int first = params[encrypt ? 0 : 1];
                final int second = params[encrypt ? 1 : 0];
                ch = StringOps.swap(ch, first, second);
            } else if (operation.startsWith("swap letter ")) {
                final String[] s =
                        operation.substring("swap letter ".length()).split(" with letter ");
                assert s[0].length() == 1 && s[1].length() == 1;
                final char first = s[encrypt ? 0 : 1].charAt(0);
                final char second = s[encrypt ? 1 : 0].charAt(0);
                ch = StringOps.swap(ch, first, second);
            } else if (operation.startsWith("reverse positions ")) {
                final Integer[] params = StringOps.getDigits(operation, 2);
                reverse(ch, params[0], params[1] + 1);
            } else if (operation.startsWith("rotate ") && !operation.startsWith("rotate based ")) {
                final Integer param = StringOps.getDigits(operation, 1)[0];
                final int amount;
                if (operation.contains("left")) {
                    amount = ch.length - (encrypt ? param : ch.length - param);
                } else {
                    amount = encrypt ? param : ch.length - param;
                }
                ch = StringOps.rotateRight(ch, amount);
            } else if (operation.startsWith("move ")) {
                final Integer[] params = StringOps.getDigits(operation, 2);
                final Integer from = params[encrypt ? 0 : 1];
                final Integer to = params[encrypt ? 1 : 0];
                ch = StringOps.move(ch, from, to);
            } else {
                final String param =
                        operation.substring(operation.length() - 1, operation.length());
                final char letter = param.charAt(0);
                if (encrypt) {
                    ch = rotateByLetter(ch, letter);
                } else {
                    char[] check1 = ch;
                    char[] check2;
                    do {
                        check1 = StringOps.rotateLeft(check1, 1);
                        check2 = rotateByLetter(check1, letter);
                    } while (!Arrays.equals(check2, ch));
                    ch = check1;
                }
            }
        }
        return String.valueOf(ch);
    }

    private char[] rotateByLetter(final char[] ch, final char c) {
        final int index = indexOf(ch, c);
        final int amount = index + 1 + (index >= 4 ? 1 : 0);
        return StringOps.rotateRight(ch, amount);
    }

    @Override
    public String solvePart1(final List<String> operations) {
        return solve(operations, "abcdefgh", TRUE);
    }

    @Override
    public String solvePart2(final List<String> operations) {
        return solve(operations, "fbgdceah", FALSE);
    }

    @Override
    @SuppressWarnings("PMD.LiteralsFirstInComparisons")
    protected void samples() {
        final AoC2016_21 test = createDebug();
        final List<String> input = test.parseInput(StringOps.splitLines(TEST));
        assert test.solve(input, "abcde", true).equals("decab");
        assert test.solve(input, "decab", false).equals("abcde");
    }

    public static void main(final String[] args) throws Exception {
        create().run();
    }

    private static final String TEST =
            """
            swap position 4 with position 0
            swap letter d with letter b
            reverse positions 0 through 4
            rotate left 1 step
            move position 1 to position 4
            move position 3 to position 0
            rotate based on position of letter b
            rotate based on position of letter d
            """;
}

File: src/main/python/AoC2016_21.py
#! /usr/bin/env python3
#
# Advent of Code 2016 Day 21
#

from aoc import my_aocd
from aoc.common import log


def _get_digits(s: str, expected: int) -> tuple[int]:
    digits = tuple(map(lambda x: int(x), filter(lambda x: x.isdigit(), s)))
    if len(digits) != expected:
        raise ValueError(f"Expected {expected}, got {len(digits)}")
    return digits


def _swap_positions(s: str, first: int, second: int) -> str:
    ch = list(s)
    ch[first], ch[second] = ch[second], ch[first]
    return "".join(ch)


def _swap_letters(s: str, first: str, second: str):
    i1 = s.index(first)
    i2 = s.index(second)
    return _swap_positions(s, i1, i2)


def _rotate_right(s: str, amount: int) -> str:
    amount = amount % len(s)
    return s[-amount:] + s[:-amount]


def _rotate_by_letter(s: str, c: str) -> str:
    index = s.index(c)
    amount = index + 1 + (1 if index >= 4 else 0)
    return _rotate_right(s, amount)


def _solve(operations: tuple[str], start: str, encrypt: bool) -> str:
    ch = start
    for i in range(len(operations)):
        idx = i if encrypt else len(operations) - 1 - i
        op = operations[idx]
        log(op)
        if op.startswith("swap position "):
            params = _get_digits(op, 2)
            first = params[0] if encrypt else params[1]
            second = params[1] if encrypt else params[0]
            ch = _swap_positions(ch, first, second)
        elif op.startswith("swap letter "):
            s = op[len("swap letter "):].split(" with letter ")
            assert len(s[0]) == 1 and len(s[1]) == 1
            first = s[0] if encrypt else s[1]
            second = s[1] if encrypt else s[0]
            ch = _swap_letters(ch, first, second)
        elif op.startswith("reverse positions "):
            params = _get_digits(op, 2)
            ch = ch[:params[0]] \
                + ch[params[0]:params[1]+1][::-1] \
                + ch[params[1]+1:]
        elif op.startswith("rotate ") and not op.startswith("rotate based "):
            param = _get_digits(op, 1)[0]
            if "left" in op:
                amount = len(ch) - (param if encrypt else len(ch) - param)
            else:
                amount = param if encrypt else len(ch) - param
            ch = _rotate_right(ch, amount)
        elif op.startswith("move "):
            params = _get_digits(op, 2)
            from_ = params[0 if encrypt else 1]
            to = params[1 if encrypt else 0]
            if from_ == to:
                raise ValueError("Expected to and from to be different")
            if from_ < to:
                ch = ch[0:from_] + ch[from_+1:to+1] + ch[from_] + ch[to+1:]
            else:
                ch = ch[0:to] + ch[from_] + ch[to:from_] + ch[from_+1:]
        elif op.startswith("rotate based "):
            letter = op[-1:]
            if encrypt:
                ch = _rotate_by_letter(ch, letter)
            else:
                ch1 = ch
                ch2 = ""
                while ch2 != ch:
                    ch1 = _rotate_right(ch1, len(ch1) - 1)
                    ch2 = _rotate_by_letter(ch1, letter)
                ch = ch1
        else:
            raise ValueError("Invalid input")
        log(ch)
    return ch


def part_1(inputs: tuple[str]) -> str:
    return _solve(inputs, "abcdefgh", True)


def part_2(inputs: tuple[str]) -> str:
    return _solve(inputs, "fbgdceah", False)


TEST = '''\
swap position 4 with position 0
swap letter d with letter b
reverse positions 0 through 4
rotate left 1 step
move position 1 to position 4
move position 3 to position 0
rotate based on position of letter b
rotate based on position of letter d
'''.splitlines()


def main() -> None:
    my_aocd.print_header(2016, 21)

    assert _solve(TEST, "abcde", True) == "decab"
    assert _solve(TEST, "decab", False) == "abcde"

    inputs = my_aocd.get_input(2016, 21, 100)
    result1 = part_1(inputs)
    print(f"Part 1: {result1}")
    result2 = part_2(inputs)
    print(f"Part 2: {result2}")


if __name__ == '__main__':
    main()
